README.md
# PyBambooHR

This is a forked version of the unofficial Python API for Bamboo HR that is compatible with Google App Engine (GAE). So far it is focusing on managing employee information but you can pretty much do anything you want with a little python.

~~The library makes use of the [requests](http://docs.python-requests.org/en/latest/) library for Python and [HTTPretty](https://github.com/gabrielfalcao/HTTPretty) for testing. A huge thank you to both of those excellent projects.~~

This forked version of the package requires using a downgraded version of requests (0.13.2) so it does not conflict with Google App Engine. Instead, it uses GAE's urlfetch(https://cloud.google.com/appengine/docs/python/urlfetch/).

To run the tests, you are required to update the tests/settings.py or create a tests/local\_settings.py file (better idea) with the GAE\_PATH setting.

Using this library is very simple:

```python
from PyBambooHR import PyBambooHR

bamboo = PyBambooHR(subdomain='yoursub', api_key='yourapikeyhere')

employees = bamboo.get_employee_directory()
```

(Note that you have to enable sharing employee directory to use that method.)

This will give you a list of employees with properties on each including their ID.


```python
from PyBambooHR import PyBambooHR

bamboo = PyBambooHR(subdomain='yoursub', api_key='yourapikeyhere')

# Jim's employee ID is 123 and we are not specifying fields so this will get all of them.
jim = bamboo.get_employee(123)

# Pam's employee ID is 222 and we are specifying fields so this will get only the ones we request.
pam = bamboo.get_employee(222, ['city', 'workPhone', 'workEmail'])

```

Adding an employee

```python
from PyBambooHR import PyBambooHR

bamboo = PyBambooHR(subdomain='yoursub', api_key='yourapikeyhere')

# The firstName and lastName keys are required...
employee = {'firstName': 'Test', 'lastName': 'Person'}

result = bamboo.add_employee(employee)

The result dict will contain id and location. "id" is the numerical BambooHR employee ID. Location is a link to that employee.

```

Updating an employee

```python
from PyBambooHR import PyBambooHR

bamboo = PyBambooHR(subdomain='yoursub', api_key='yourapikeyhere')

# His name was test person...
employee = {'firstName': 'Another', 'lastName': 'Namenow'}

# Use the ID and then the dict with the new information
result = bamboo.update_employee(333, employee)

result will be True or False depending on if it succeeded.

```

Requesting a Report

```python
from PyBambooHR import PyBambooHR

bamboo = PyBambooHR(subdomain='yoursub', api_key='yourapikeyhere')

# Use the ID to request json information
result = bamboo.request_company_report(1, format='json', filter_duplicates=True)

# Now do stuff with your results (Will vary by report.)
for employee in result['employees']:
    print(employee)

# Use the ID and save a pdf:
result = bamboo.request_company_report(1, format='pdf', output_file='/tmp/report.pdf', filter_duplicates=True)

```


CHANGES
CHANGES
=======
Version 0.7.0 - 18 April 2016
  * Convert to use Google App Engine's (GAE) urlfetch instead of latest requests
    library. Requires reverting to requests version 0.13.2 for GAE compatibility.

Version 0.4.2 - 03 October 2014

  * Added row_id to row dictionary to keep track of them in case we need them.

Version 0.3.6 - 03 October 2014

  * Improvements by Ng Zhi An

Version 0.3.1 - 24 February 2014

  * Raise errors on non-200 statuses

Version 0.3.0 - 23 February 2014

  * Company Reports
  * Custom Reports

Version 0.2.0 - 05 October 2013

  * Underscore key support
  * Better test coverage

Version 0.1.0 - 02 October 2013

  * Pre-Alpha version
  * Uses Bamboo HR API to get employee info.


tests/settings.py
# Put in local_settings.py so you can exclude from upload.
# example: '/home/[username]/google-cloud-sdk/platform/google_appengine'
GAE_PATH = None

try:
    from local_settings import *
except ImportError:
    pass


tests/test_misc.py
#!/usr/bin/env python
#encoding:utf-8
#author:smeggingsmegger/Scott Blevins
#project:PyBambooHR
#repository:http://github.com/smeggingsmegger/PyBambooHR
#license:agpl-3.0 (http://www.gnu.org/licenses/agpl-3.0.en.html)

"""Unittests for misc. functions
"""

import os
import sys
import unittest

# Force parent directory onto path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Insert GAE
import settings
sys.path.insert(0, settings.GAE_PATH)

from PyBambooHR import PyBambooHR, utils


class test_misc(unittest.TestCase):
    # Used to store the cached instance of PyBambooHR
    bamboo = None

    def setUp(self):
        if self.bamboo is None:
            self.bamboo = PyBambooHR(subdomain='test', api_key='testingnotrealapikey')

    def test_employee_xml(self):
        employee = {
            'firstName': 'Test',
            'lastName': 'Person'
        }
        xml = self.bamboo._format_employee_xml(employee)
        self.assertIn('<field id="firstName">Test</field>', xml)
        self.assertIn('<field id="lastName">Person</field>', xml)

    def test_init_value_errors(self):
        self.assertRaises(ValueError, PyBambooHR, {'subdomain': 'test'})
        self.assertRaises(ValueError, PyBambooHR, {'api_key': 'testingnotrealapikey'})
        self.assertIsNotNone(PyBambooHR(subdomain='test', api_key='testingnotrealapikey'))

    def test_xml(self):
        xml = """<?xml version="1.0"?>
                     <table>
                       <row id="321" employeeId="123">
                           <field id="customFieldA">123 Value A</field>
                           <field id="customFieldB">123 Value B</field>
                       </row>
                       <row id="999" employeeId="321">
                           <field id="customFieldA">321 Value A</field>
                           <field id="customFieldB">321 Value B</field>
                       </row>
                     </table>"""
        obj = utils._parse_xml(xml)

        row_id_one = obj['table']['row'][0]['@id']
        row_id_two = obj['table']['row'][1]['@id']

        self.assertEqual('321', row_id_one)
        self.assertEqual('999', row_id_two)

        employee_id_one = obj['table']['row'][0]['@employeeId']
        employee_id_two = obj['table']['row'][1]['@employeeId']

        self.assertEqual('123', employee_id_one)
        self.assertEqual('321', employee_id_two)

        rows = utils._extract(obj, 'table', 'row')

        self.assertEqual('123 Value A', rows[0]['field'][0]['#text'])
        self.assertEqual('123 Value B', rows[0]['field'][1]['#text'])

        self.assertEqual('321 Value A', rows[1]['field'][0]['#text'])
        self.assertEqual('321 Value B', rows[1]['field'][1]['#text'])

    def test_make_field_xml(self):
        xml = utils.make_field_xml('123', 'T&C')
        self.assertEqual('<field id="123">T&amp;C</field>', xml)

        xml = utils.make_field_xml('123', 'T&C', pre='\t', post='\n')
        self.assertEqual('\t<field id="123">T&amp;C</field>\n', xml)

        xml = utils.make_field_xml('123', None, pre='\t', post='\n')
        self.assertEqual('\t<field id="123" />\n', xml)
        pass

    def test__format_row_xml(self):
        row = {'f1': 'v1', 'f2': 'v2'}
        xml = self.bamboo._format_row_xml(row)
        self.assertIn('<field id="f1">v1</field>', xml)
        self.assertIn('<field id="f2">v2</field>', xml)

    def test_transform_tabular_data(self):
        xml = """<?xml version="1.0"?>
                 <table>
                   <row id="321" employeeId="123">
                     <field id="customFieldA">123 Value A</field>
                     <field id="customFieldB">123 Value B</field>
                   </row>
                   <row id="999" employeeId="321">
                     <field id="customFieldA">321 &amp; Value A</field>
                   </row>
                 </table>"""
        table = {'123': [{
                         'customFieldA': '123 Value A',
                         'customFieldB': '123 Value B',
                         'row_id': '321'}],
                 '321': [{
                         'customFieldA': '321 & Value A',
                         'row_id': '999'}]}
        self.assertEqual(table, utils.transform_tabular_data(xml))

    def test_transform_tabular_data_single_row(self):
        xml = """<?xml version="1.0"?>
                 <table>
                   <row id="321" employeeId="123">
                     <field id="customFieldA">123 Value A</field>
                   </row>
                 </table>"""
        table = {'123': [{'customFieldA': '123 Value A', 'row_id': '321'}]}
        self.assertEqual(table, utils.transform_tabular_data(xml))

    def test_transform_tabular_data_empty_table(self):
        xml = """<?xml version="1.0"?>
                     <table/>"""
        table = {}
        self.assertEqual(table, utils.transform_tabular_data(xml))

    def test_transform_tabular_data_empty_field(self):
        xml = """<?xml version="1.0"?>
                 <table>
                   <row id="321" employeeId="123">
                     <field id="customFieldA">123 Value A</field>
                     <field id="customFieldC"></field>
                   </row>
                   <row id="999" employeeId="321">
                     <field id="customFieldB">321 Value B</field>
                   </row>
                 </table>"""
        table = {'123': [{'customFieldA': '123 Value A',
                          'customFieldC': None,
                          'row_id': '321'}],
                 '321': [{'customFieldB': '321 Value B',
                          'row_id': '999'}]}

        self.assertEqual(table, utils.transform_tabular_data(xml))


tests/test_reports.py
#!/usr/bin/env python
#encoding:utf-8
#author:smeggingsmegger/Scott Blevins
#project:PyBambooHR
#repository:http://github.com/smeggingsmegger/PyBambooHR
#license:agpl-3.0 (http://www.gnu.org/licenses/agpl-3.0.en.html)

"""Unittests for employees api
"""

import httpretty
import os
import sys
import unittest

from json import dumps

# Force parent directory onto path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Insert GAE
import settings
sys.path.insert(0, settings.GAE_PATH)
sys.path.insert(0, settings.GAE_PATH+'/lib/yaml/lib')
from google.appengine.ext import testbed

from PyBambooHR import PyBambooHR

class test_reports(unittest.TestCase):
    # Used to store the cached instance of PyBambooHR
    bamboo = None

    # Another instance, using underscore keys
    bamboo_u = None

    # Common report response body:
    body = None

    def setUp(self):
        self.testbed = testbed.Testbed()
        self.testbed.activate()
        self.testbed.init_urlfetch_stub()
        
        if self.body is None:
            self.body = dumps({"title":"Template","fields":[{"id":"fullName2","type":"text","name":"Last Name, First Name"},{"id":"employeeNumber","type":"employee_number","name":"Employee #"},{"id":"hireDate","type":"date","name":"Hire date"}],"employees":[{"id":"123","fullName2":"Person, Test","employeeNumber":"00001","hireDate":"2010-12-15"},{"id":"124","fullName2":"Guy, Someother","employeeNumber":"00002","hireDate":"2008-10-13"},{"id":"125","fullName2":"Here, Someone","employeeNumber":"00003","hireDate":"2011-03-04"}]})

        if self.bamboo is None:
            self.bamboo = PyBambooHR(subdomain='test', api_key='testingnotrealapikey')

        if self.bamboo_u is None:
            self.bamboo_u = PyBambooHR(subdomain='test', api_key='testingnotrealapikey', underscore_keys=True)

    @httpretty.activate
    def test_request_company_report(self):
        httpretty.register_uri(httpretty.GET, "https://api.bamboohr.com/api/gateway.php/test/v1/reports/1?format=json&fd=yes",
                               status=200, body=self.body, content_type="application/json")

        result = self.bamboo.request_company_report(1, report_format='json', filter_duplicates=True)
        self.assertIsNotNone(result['fields'])
        self.assertEquals('123', result['employees'][0]['id'])

    @httpretty.activate
    def test_company_report_format_failure(self):
        httpretty.register_uri(httpretty.GET, "https://api.bamboohr.com/api/gateway.php/test/v1/reports/1?format=json&fd=yes",
                               status=200, body=self.body, content_type="application/json")

        self.assertRaises(UserWarning, self.bamboo.request_company_report, 1, report_format='gif', filter_duplicates=True)

    @httpretty.activate
    def test_request_custom_report(self):
        httpretty.register_uri(httpretty.POST, "https://api.bamboohr.com/api/gateway.php/test/v1/reports/custom/?format=xls",
                               status=200, body=self.body, content_type="application/vnd.ms-excel")

        result = self.bamboo.request_custom_report(['id', 'firstName', 'lastName', 'workEmail'], report_format='xls')
        self.assertEquals(result.headers['status'], '200')
        self.assertEquals(result.headers['content-type'], 'application/vnd.ms-excel')

    @httpretty.activate
    def test_company_custom_format_failure(self):
        httpretty.register_uri(httpretty.GET, "https://api.bamboohr.com/api/gateway.php/test/v1/reports/1?format=json&fd=yes",
                               status=200, body=self.body, content_type="application/json")

        self.assertRaises(UserWarning, self.bamboo.request_custom_report, 1, report_format='gif')



PyBambooHR/utils.py
"""
A collection of misc. utilities that are used in the main class.
"""

import datetime
import re
import xmltodict
import urllib

from requests import HTTPError
from google.appengine.api import urlfetch

def camelcase_keys(data):
    """
    Converts all the keys in a dict to camelcase. It works recursively to convert any nested dicts as well.
    @param data: The dict to convert
    """
    return_dict = {}
    for key in data:
        if isinstance(data[key], dict):
            return_dict[underscore_to_camelcase(key)] = camelcase_keys(data[key])
        else:
            return_dict[underscore_to_camelcase(key)] = data[key]

    return return_dict

def camelcase_to_underscore(name):
    """
    Converts a string to underscore. (Typically from camelcase.)
    @param name: The string to convert.
    """
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)).lower()

def underscore_to_camelcase(name):
    """
    Converts a string to camelcase. (Typically from underscore.)
    @param name: The string to convert.
    """
    return re.sub(r'_([a-z])', lambda m: (m.group(1).upper()), name)

def underscore_keys(data):
    """
    Converts all the keys in a dict to camelcase. It works recursively to convert any nested dicts as well.
    @param data: The dict to convert
    """
    return_dict = {}
    for key in data:
        if isinstance(data[key], dict):
            return_dict[camelcase_to_underscore(key)] = underscore_keys(data[key])
        else:
            return_dict[camelcase_to_underscore(key)] = data[key]

    return return_dict

_date_regex = re.compile(r"^\d{4}-\d{2}-\d{2}")


def make_field_xml(id, value=None, pre='', post=''):
    id = escape(str(id))
    if value:
        value = escape(str(value))
        tag = '<field id="{}">{}</field>'.format(id, value)
    else:
        tag = '<field id="{}" />'.format(id)
    return '{0}{1}{2}'.format(pre, tag, post)


def resolve_date_argument(arg):
    if isinstance(arg, (datetime.datetime, datetime.date)):
        return arg.strftime('%Y-%m-%d')
    elif isinstance(arg, basestring) and _date_regex.match(arg):
        return arg
    elif arg is None:
        return None
    raise ValueError("Date argument {} must be either datetime, date, or string in form YYYY-MM-DD".format(arg))

def transform_tabular_data(xml_input):
    """
    Converts table data (xml) from BambooHR into a dictionary with employee
    id as key and a list of dictionaries.
    Each field is a dict with the id as the key and inner text as the value
    e.g.
        <table>
          <row id="321" employeeId="123">
            <field id="customFieldA">123 Value A</field>
            <field id="customFieldC"></field>
          </row>
          <row id="999" employeeId="321">
            <field id="customFieldB">321 Value B</field>
          </row>
        </table>
    becomes
        {'123': [{
                 'customFieldA': '123 Value A',
                 'customFieldB': '123 Value B',
                 'row_id': '321'}],
         '321': [{
                 'customFieldA': '321 Value A',
                 'row_id': '999'}]}
    """
    obj = _parse_xml(xml_input)
    rows = _extract(obj, 'table', 'row')
    by_employee_id = {}
    for row in rows:
        eid = row['@employeeId']
        field_list = row['field'] if type(row['field']) is list \
            else [row['field']]
        fields = dict([(f['@id'], f.get('#text', None)) for f in field_list])
        fields['row_id'] = row['@id']
        by_employee_id.setdefault(eid, []).append(fields)
    return by_employee_id

def transform_whos_out(xml_input):
    obj = _parse_xml(xml_input)
    rows = _extract(obj, 'calendar', 'item')
    events = []
    for row in rows:
        ev = {
            'type': row['@type'],
            'start': row['start'],
            'end': row['end']
        }
        if ev['type'] == 'timeOff':
            ev['employeeId'] = row['employee']['@id']
            ev['employeeName'] = row['employee']['#text']
        events.append(ev)
    return events

def transform_time_off(xml_input):
    obj = _parse_xml(xml_input)
    rows = _extract(obj, 'requests', 'request')
    requests = []
    for row in rows:
        # Sample XML
        # <employee id="1">Jon Doe</employee>
        # <status lastChanged="2011-08-14" lastChangedByUserId="1">approved</status>
        # <start>2001-01-01</start>
        # <end>2001-01-06</end>
        # <created>2011-08-13</created>
        # <type id="1">Vacation</type>
        # <amount unit="days">5</amount>
        rq = {
            'employeeId': row['employee']['@id'],
            'employeeName': row['employee']['#text'],
            'status': row['status']['#text'],
            'type': row['type']['#text'],
            'amount': row['amount']['#text'],
            'unit': row['amount']['@unit'],
            'start': row['start'],
            'end': row['end']
        }
        requests.append(rq)
    return requests

def transform_change_list(xml_input):
    obj = _parse_xml(xml_input)
    rows = _extract(obj, 'changeList', 'employee')
    events = []
    for row in rows:
        events.append({
            'id': row['@id'],
            'action': row['@action'],
            'lastChanged': datetime.datetime.strptime(row['@lastChanged'], '%Y-%m-%dT%H:%M:%S+00:00')
        })
    return events

def _extract(xml_obj, first_key, second_key):
    first = xml_obj.get(first_key, {}) or {}
    rows = first.get(second_key, []) or []
    return rows if isinstance(rows, list) else [rows]

def _parse_xml(input):
    return xmltodict.parse(input)


XML_ESCAPES = (
    ('<', '&lt;'),
    ('>', '&gt;'),
    ('&', '&amp;'),
    ("'", '&apos;'),
    ('"', '&quot;'),
)


def escape(to_escape):
    """Returns the given string with XML reserved characters encoded."""
    for char, repl in XML_ESCAPES:
        to_escape = to_escape.replace(char, repl)
    return to_escape


def urlfetch_gae(url, payload=None, method=1, headers={}, allow_truncated=False, follow_redirects=True, deadline=None, validate_certificate=None):
    """Uses GAE's urlfetch in order to fetch a URL. For method arg, use urlfetch.GET, urlfetch.POST, etc."""
    http_error_msg = None
    try:
        if payload:
            # BambooHR wants it in the URL
            try:
                url = url+'?'+urllib.urlencode(payload)
            except TypeError:
                pass
        
        response = urlfetch.fetch(url=url, payload=payload, method=method, headers=headers,
                                  allow_truncated=allow_truncated, follow_redirects=follow_redirects,
                                  deadline=deadline, validate_certificate=validate_certificate)

        if 400 <= response.status_code < 500:
            http_error_msg = '%s Client Error for url: %s' % (response.status_code, url)

        elif 500 <= response.status_code < 600:
            http_error_msg = '%s Server Error for url: %s' % (response.status_code, url)

    except urlfetch.InvalidURLError:
        http_error_msg =  "URL is an empty string or obviously invalid"

    except urlfetch.DownloadError:
        http_error_msg = "Server cannot be contacted"

    if http_error_msg:
        raise HTTPError(http_error_msg)
        
    return response
